src/Attributes.php

```diff
@@ -0,0 +1,142 @@
+     *
+     * @param Escaper $htmlEscaper HTML escaper
+     * @param Escaper $htmlAttrEscaper Attribute escaper
+     * @param array|Traversable $attribs Attributes
+     */
+    public function __construct($htmlEscaper, $htmlAttrEscaper, $attribs = [])
+    {
+        parent::__construct();
+        $this->htmlEscaper = $htmlEscaper;
+        $this->htmlAttrEscaper = $htmlAttrEscaper;
+        foreach ($attribs as $name => $value) {
+            $this->offsetSet($name, $value);
+        }
+    }
+
+    /**
+     * Add a value to an attribute.
+     *
+     * Sets the attribute if it does not exist.
+     *
+     * @param $name string Name
+     * @param $value string|array Value
+     *
+     * @return Attributes
+     */
+    public function add($name, $value)
+    {
+        if ($this->offsetExists($name)) {
+            $this->offsetSet($name, array_merge((array) $this->offsetGet($name), (array) $value));
+        } else {
+            $this->offsetSet($name, $value);
+        }
+        return $this;
+    }
+
+    /**
+     * Merge attributes with existing attributes.
+     *
+     * @param $attribs array|Traversable Attributes
+     *
+     * @return $this
+     */
+    public function merge($attribs)
+    {
+        foreach ($attribs as $name => $value) {
+            $this->add($name, $value);
+        }
+        return $this;
+    }
+
+    /**
+     * Does a specific attribute with a specific value exist?
+     *
+     * @param $name string Name
+     * @param $value string Value
+     *
+     * @return bool
+     */
+    public function hasValue($name, $value)
+    {
+        if ($this->offsetExists($name)) {
+            $storeValue = $this->offsetGet($name);
+            if (is_array($storeValue)) {
+                return in_array($value, $storeValue);
+            } else {
+                return $value === $storeValue;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Return a string of tag attributes.
+     *
+     * @return string
+     */
+    public function __toString()
+    {
+        $xhtml          = '';
+
+        foreach ($this->getArrayCopy() as $key => $val) {
+            $key = $this->htmlEscaper->escapeHtml($key);
+
+            if (0 === strpos($key, 'on') || ('constraints' == $key)) {
+                // Don't escape event attributes; _do_ substitute double quotes with singles
+                if (! is_scalar($val)) {
+                    // non-scalar data should be cast to JSON first
+                    $val = \Laminas\Json\Json::encode($val);
+                }
+            } else {
+                if (is_array($val)) {
+                    $val = implode(' ', $val);
+                }
+            }
+
+            $val = $this->htmlAttrEscaper->escapeHtmlAttr($val);
+
+            if (strpos($val, '"') !== false) {
+                $xhtml .= " $key='$val'";
+            } else {
+                $xhtml .= " $key=\"$val\"";
+            }
+        }
+
+        return $xhtml;
```

Comment:
```suggestion
        $xhtml = '';

        foreach ($this->getArrayCopy() as $key => $val) {
            $key = $this->htmlEscaper->escapeHtml($key);

            if ((0 === strpos($key, 'on') || ('constraints' === $key)) && ! is_scalar($val)) {
                // Don't escape event attributes; _do_ substitute double quotes with singles
                // non-scalar data should be cast to JSON first
                $val = json_encode($val, JSON_HEX_TAG | JSON_HEX_APOS | JSON_HEX_QUOT | JSON_HEX_AMP);
            }

            if (0 !== strpos($key, 'on') && 'constraints' !== $key && is_array($val)) {
                // Non-event keys and non-constraints keys with array values
                // should have values separated by whitespace
                $val = implode(' ', $val);
            }

            $val    = $this->htmlAttrEscaper->escapeHtmlAttr($val);
            $quote  = strpos($value, '"') !== false ? "'" : '"';
            $xhtml .= sprintf(' %2$s=%1$s%3$s%1$s', $quote, $key, $value);
        }

        return $xhtml;
```

Notes:

- laminas-json is in the process of being deprecated, as it doesn't really provide any benefits over ext-json
- Ideally, I'd extract the parts where you massage the `$val` into a method, so you could return early. However, two conditionals in sequence are better than nested conditionals.
- I use positional placeholders with `sprintf` to perform quoting (e.g. `%1$s` says to inject argument 1 as a string; I do this so we can re-use the `$quote` character we've previously discovered); variable interpolation is largely considered a no-no within Laminas components.